File: compiler/analysis/parser/parser.go
package parser

import (
	"mangrove/compiler/analysis"
	"mangrove/compiler/analysis/token"
)

type Parser struct {
	tokens       []token.Token
	position     int
	readPosition int
	tok          token.Token
}

func CreateNewParser(lexer *analysis.Lexer) *Parser {
	var tokens []token.Token
	var tok token.Token

	for {
		tok = lexer.NextToken()

		if tok.Type == token.EOF {
			break
		} else {
			tokens = append(tokens, tok)
		}
	}

	return &Parser{
		tokens: tokens,
	}
}

func (parser *Parser) PeekNextToken() token.Token {
	return parser.tokens[parser.readPosition]
}

func (parser *Parser) NextToken() {
	if parser.readPosition >= len(parser.tokens) {
		parser.tok = token.Token{Type: token.EOF}
	} else {
		parser.tok = parser.tokens[parser.readPosition]
	}

	parser.position = parser.readPosition
	parser.readPosition += 1
}

func (parser *Parser) NextExpression() Expression {
	var expression Expression
	var tok token.Token
	var subexpressionPossible bool

	tok = parser.PeekNextToken()

	if tok.Type == token.EOF {
		break
	}

	switch tok.Type {
	case token.TYPE_UINT8 :
		expression = &NumberExpession{Token: tok }
	case token.TYPE_UINT16:
		expression = &NumberExpession{Token: tok }
	case token.TYPE_UINT32:
		expression = &NumberExpession{Token: tok }
	case token.TYPE_UINT64:
		expression = &NumberExpession{Token: tok }
	case token.TYPE_INT8  :
		expression = &NumberExpession{Token: tok }
	case token.TYPE_INT16 :
		expression = &NumberExpession{Token: tok }
	case token.TYPE_INT32 :
		expression = &NumberExpession{Token: tok }
	case token.TYPE_INT64 :
		expression = &NumberExpession{Token: tok }
	case token.VARIABLE:
		expression = parser.ParseVariableDeclarationExpression()
	}

	return expression
}

File: Mangrove.Compiler/SyntaxAnalysis/Parser.cs
namespace Mangrove.Compiler.SyntaxAnalysis;

public class Parser {
    public Token[] InputTokens;
    public int     CurrentPosition;
    public int     ReadPosition;
    public Token   CurrentToken;

    public Parser(IEnumerable<Token> tokens) {
        this.InputTokens = tokens.ToArray();
    }

    public void ReadToken() {
        if (this.ReadPosition >= this.InputTokens.Length) {
            this.CurrentToken = new Token { TokenType = TokenType.EndOfFile };
        } else {
            this.CurrentToken = this.InputTokens[this.ReadPosition];
        }

        this.CurrentPosition = this.ReadPosition;
        this.ReadPosition++;
    }

    public Token PeekToken() {
        if (this.ReadPosition >= this.InputTokens.Length) {
            return new Token { TokenType = TokenType.EndOfFile };
        }

        return this.InputTokens[this.ReadPosition];
    }

    public ExpressionValue.Variable ParseVariable() {
        ExpressionValue.Variable returnVariable;

        this.ReadToken(); //var

        if (this.CurrentToken.TokenType != TokenType.GreaterThan) {
            throw new ExpectedTypeParameterOnVariable(this.CurrentToken.Line, this.CurrentToken.Column);
        }

        this.ReadToken(); //<

        Token typeLiteral = this.CurrentToken;

        Dictionary<TokenType, bool> allowedTypesAndConversion = new Dictionary<TokenType, bool>() {
            [TokenType.Identifier] = true, [TokenType.TypeInt8]   = true, [TokenType.TypeInt16]  = true,
            [TokenType.TypeInt32]  = true, [TokenType.TypeInt64]  = true, [TokenType.TypeUint8]  = true,
            [TokenType.TypeUint16] = true, [TokenType.TypeUint32] = true, [TokenType.TypeUint64] = true,
            [TokenType.TypeDouble] = true, [TokenType.TypeFloat]  = true, [TokenType.TypeString] = true,
            [TokenType.Array]      = true
        };

        if (!allowedTypesAndConversion.ContainsKey(typeLiteral.TokenType)) {
            throw new InvalidVariableType(typeLiteral.Line, typeLiteral.Column, typeLiteral);
        }

        this.ReadToken(); //>

        Token nameLiteral = this.CurrentToken;

        if (nameLiteral.TokenType != TokenType.Identifier) {
            throw new ExpectedNameOnVariable(this.CurrentToken.Line, this.CurrentToken.Column);
        }

        this.ReadToken();
    }

    public Expression ParseNextExpression(ExpressionOperatorPrescedence prescedence = ExpressionOperatorPrescedence.Lowest) {
        Expression expression = default;

        switch (this.CurrentToken.TokenType) {
            case TokenType.Variable:
                ExpressionValue.Variable variable = this.ParseVariable();
                break;
        }

        while (true) {
            this.ReadToken();

            if (this.CurrentToken.TokenType == TokenType.EndOfFile) {
                break;
            }

            Token currentToken = this.CurrentToken;

            if (currentToken.IsOperator) {
                ExpressionOperator expressionOperator = ExpressionOperator.FromTokenType(currentToken.TokenType);

                if ((expressionOperator.Type & ExpressionOperatorType.Infix) == 0)
                    throw new InvalidInfixOperator(this.CurrentToken.Line, this.CurrentToken.Column, this.CurrentToken);

                if (prescedence >= expressionOperator.Prescedence)
                    break;

                this.ReadToken();


                expression = new Expression(
                    new ExpressionValue.InfixExpression(
                        expressionOperator,
                        expression,
                        this.ParseNextExpression(expressionOperator.Prescedence)
                        )
                    );
            } else {

            }
        }
    }
}
